components/script/dom/vr.rs

```diff
@@ -0,0 +1,173 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use dom::bindings::cell::DOMRefCell;
+use dom::bindings::codegen::Bindings::VRBinding;
+use dom::bindings::codegen::Bindings::VRBinding::VRMethods;
+use dom::bindings::error::Error;
+use dom::bindings::inheritance::Castable;
+use dom::bindings::js::{JS, Root};
+use dom::bindings::reflector::{DomObject, reflect_dom_object};
+use dom::event::Event;
+use dom::eventtarget::EventTarget;
+use dom::globalscope::GlobalScope;
+use dom::promise::Promise;
+use dom::vrdisplay::VRDisplay;
+use dom::vrdisplayevent::VRDisplayEvent;
+use ipc_channel::ipc;
+use ipc_channel::ipc::IpcSender;
+use script_traits::WebVREventMsg;
+use std::rc::Rc;
+use vr_traits::WebVRMsg;
+use vr_traits::webvr;
+
+#[dom_struct]
+pub struct VR {
+    eventtarget: EventTarget,
+    displays: DOMRefCell<Vec<JS<VRDisplay>>>
+}
+
+impl VR {
+    fn new_inherited() -> VR {
+        VR {
+            eventtarget: EventTarget::new_inherited(),
+            displays: DOMRefCell::new(Vec::new())
+        }
+    }
+
+    pub fn new(global: &GlobalScope) -> Root<VR> {
+        let root = reflect_dom_object(box VR::new_inherited(),
+                           global,
+                           VRBinding::Wrap);
+        root.register();
+        root
+    }
+}
+
+impl Drop for VR {
+    fn drop(&mut self) {
+        self.unregister();
+    }
+}
+
+impl VRMethods for VR {
+    #[allow(unrooted_must_root)]
+    // https://w3c.github.io/webvr/#interface-navigator
+    fn GetDisplays(&self) -> Rc<Promise> {
+        let promise = Promise::new(&self.global());
+        if !self.VrEnabled() {
+            // WebVR spec: The Promise MUST be rejected if the Document object is inside
+            // an iframe that does not have the allowvr attribute set.
+            promise.reject_error(promise.global().get_cx(), Error::Security);
+            return promise;
+        }
+
+        if let Some(wevbr_sender) = self.webvr_thread() {
```

Comment:
Spelling of `webvr_sender` is wrong, and it is weird to call the variable `sender` when the method says `thread`.